File: ros2_ws/src/siyi_camera_bridge/siyi_camera_bridge/rtsp_publisher.py
#!/usr/bin/env python3
"""
SIYI A8 Mini RTSP → ROS2 Image bridge node.

Reads RTSP stream from SIYI A8 Mini camera via OpenCV,
optionally flips the image 180° (camera mounted upside-down),
and publishes sensor_msgs/Image + sensor_msgs/CameraInfo.
"""

import os
import numpy as np

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge

# Set RTSP TCP transport BEFORE importing cv2 / opening capture
os.environ['OPENCV_FFMPEG_CAPTURE_OPTIONS'] = 'rtsp_transport;tcp'
import cv2  # noqa: E402


class RtspPublisher(Node):
    """ROS2 node that bridges an RTSP camera stream to Image + CameraInfo topics."""

    def __init__(self):
        super().__init__('siyi_rtsp_publisher')

        # ── Parameters ──────────────────────────────────────────────
        self.declare_parameter('rtsp_url',
                               'rtsp://192.168.168.16:8554/main.264')
        self.declare_parameter('frame_id', 'siyi_camera_optical_frame')
        self.declare_parameter('flip_180', False)
        self.declare_parameter('target_fps', 30.0)
        self.declare_parameter('image_width', 1280)
        self.declare_parameter('image_height', 720)
        # Fallback intrinsics matching SIYI A8 Mini HFOV ≈ 81°
        # fx = fy = (1280 / 2) / tan(81° / 2) = 640 / tan(40.5°) ≈ 749.338
        self.declare_parameter('camera_fx', 749.338)
        self.declare_parameter('camera_fy', 749.338)
        self.declare_parameter('camera_cx', 640.0)
        self.declare_parameter('camera_cy', 360.0)
        self.declare_parameter('camera_d', [0.0, 0.0, 0.0, 0.0, 0.0])

        self.rtsp_url = self.get_parameter('rtsp_url').value
        self.frame_id = self.get_parameter('frame_id').value
        self.flip_180 = self.get_parameter('flip_180').value
        self.target_fps = self.get_parameter('target_fps').value
        self.image_width = self.get_parameter('image_width').value
        self.image_height = self.get_parameter('image_height').value
        self.camera_fx = self.get_parameter('camera_fx').value
        self.camera_fy = self.get_parameter('camera_fy').value
        self.camera_cx = self.get_parameter('camera_cx').value
        self.camera_cy = self.get_parameter('camera_cy').value
        self.camera_d = self.get_parameter('camera_d').value

        # ── Publishers ──────────────────────────────────────────────
        self.image_pub = self.create_publisher(Image, '/siyi/image_raw', 10)
        self.info_pub = self.create_publisher(CameraInfo, '/siyi/camera_info', 10)

        # ── Bridge ──────────────────────────────────────────────────
        self.bridge = CvBridge()

        # ── Build static CameraInfo message ─────────────────────────
        self.camera_info_msg = self._build_camera_info()

        # ── Open RTSP capture ───────────────────────────────────────
        self.cap = None
        self._open_capture()

        # ── Timer for frame grabbing ────────────────────────────────
        timer_period = 1.0 / self.target_fps
        self.timer = self.create_timer(timer_period, self._timer_callback)

        # ── Stats ───────────────────────────────────────────────────
        self.frame_count = 0
        self.fail_count = 0
        self.max_consecutive_fails = 30  # Reconnect after this many failures

        self.get_logger().info(
            f'SIYI RTSP Publisher started: url={self.rtsp_url} '
            f'flip_180={self.flip_180} target_fps={self.target_fps} '
            f'resolution={self.image_width}x{self.image_height}'
        )

    def _open_capture(self):
        """Open or reopen the RTSP VideoCapture."""
        if self.cap is not None:
            self.cap.release()

        self.get_logger().info(f'Opening RTSP stream: {self.rtsp_url}')
        self.cap = cv2.VideoCapture(self.rtsp_url, cv2.CAP_FFMPEG)

        if not self.cap.isOpened():
            self.get_logger().error(
                f'Failed to open RTSP stream: {self.rtsp_url}. '
                'Will retry on next timer tick.'
            )
            return False

        # Try to set resolution (RTSP source may ignore this)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, self.image_width)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, self.image_height)
        # Minimize buffer to reduce latency
        self.cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)

        self.fail_count = 0
        self.get_logger().info('RTSP stream opened successfully')
        return True

    def _build_camera_info(self) -> CameraInfo:
        """Build a CameraInfo message with fallback intrinsics."""
        msg = CameraInfo()
        msg.header.frame_id = self.frame_id
        msg.width = self.image_width
        msg.height = self.image_height
        msg.distortion_model = 'plumb_bob'
        msg.d = list(self.camera_d)

        # Intrinsic camera matrix (K)
        fx = self.camera_fx
        fy = self.camera_fy
        cx = self.camera_cx
        cy = self.camera_cy
        msg.k = [
            fx,  0.0, cx,
            0.0, fy,  cy,
            0.0, 0.0, 1.0,
        ]

        # Rectification matrix (identity for monocular)
        msg.r = [
            1.0, 0.0, 0.0,
            0.0, 1.0, 0.0,
            0.0, 0.0, 1.0,
        ]

        # Projection matrix (P)
        msg.p = [
            fx,  0.0, cx,  0.0,
            0.0, fy,  cy,  0.0,
            0.0, 0.0, 1.0, 0.0,
        ]

        return msg

    def _timer_callback(self):
        """Grab a frame from RTSP and publish."""
        if self.cap is None or not self.cap.isOpened():
            self.fail_count += 1
            if self.fail_count % 30 == 1:
                self.get_logger().warn(
                    f'RTSP not open, attempting reconnect... '
                    f'(fail_count={self.fail_count})'
                )
            self._open_capture()
            return

        ret, frame = self.cap.read()

        if not ret or frame is None:
            self.fail_count += 1
            if self.fail_count >= self.max_consecutive_fails:
                self.get_logger().warn(
                    f'Lost RTSP stream after {self.fail_count} failures, reconnecting...'
                )
                self._open_capture()
            return

        self.fail_count = 0
        self.frame_count += 1

        # Flip 180° if camera is mounted upside-down
        if self.flip_180:
            frame = cv2.flip(frame, -1)

        # Timestamp
        stamp = self.get_clock().now().to_msg()

        # Publish Image
        try:
            img_msg = self.bridge.cv2_to_imgmsg(frame, encoding='bgr8')
            img_msg.header.stamp = stamp
            img_msg.header.frame_id = self.frame_id
            self.image_pub.publish(img_msg)
        except Exception as e:
            self.get_logger().error(f'cv_bridge error: {e}')
            return

        # Publish CameraInfo (same timestamp)
        self.camera_info_msg.header.stamp = stamp
        self.info_pub.publish(self.camera_info_msg)

        # Log stats periodically
        if self.frame_count % (int(self.target_fps) * 5) == 0:
            self.get_logger().info(
                f'Published {self.frame_count} frames '
                f'({frame.shape[1]}x{frame.shape[0]}) '
                f'flip_180={self.flip_180}'
            )

    def destroy_node(self):
        """Clean up on shutdown."""
        if self.cap is not None:
            self.cap.release()
            self.get_logger().info('RTSP capture released')
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = RtspPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/precision_landing/src/rtsp_publisher.cpp
#include "precision_landing/rtsp_publisher.hpp"
#include <cstdlib>
#include <chrono>

namespace precision_landing
{

RtspPublisher::RtspPublisher(const rclcpp::NodeOptions & options)
: Node("siyi_rtsp_publisher", options)
{
  // Set RTSP TCP transport BEFORE opening capture
  setenv("OPENCV_FFMPEG_CAPTURE_OPTIONS", "rtsp_transport;tcp", 1);

  // Declare Parameters
  this->declare_parameter<std::string>("rtsp_url", "rtsp://192.168.168.14:8554/main.264");
  this->declare_parameter<std::string>("frame_id", "siyi_camera_optical_frame");
  this->declare_parameter<bool>("flip_180", true);
  this->declare_parameter<double>("target_fps", 30.0);
  this->declare_parameter<int>("image_width", 1280);
  this->declare_parameter<int>("image_height", 720);
  this->declare_parameter<double>("camera_fx", 749.338);
  this->declare_parameter<double>("camera_fy", 749.338);
  this->declare_parameter<double>("camera_cx", 640.0);
  this->declare_parameter<double>("camera_cy", 360.0);
  this->declare_parameter<std::vector<double>>("camera_d", {0.0, 0.0, 0.0, 0.0, 0.0});

  // Get Parameters
  rtsp_url_ = this->get_parameter("rtsp_url").as_string();
  frame_id_ = this->get_parameter("frame_id").as_string();
  flip_180_ = this->get_parameter("flip_180").as_bool();
  target_fps_ = this->get_parameter("target_fps").as_double();
  image_width_ = this->get_parameter("image_width").as_int();
  image_height_ = this->get_parameter("image_height").as_int();
  camera_fx_ = this->get_parameter("camera_fx").as_double();
  camera_fy_ = this->get_parameter("camera_fy").as_double();
  camera_cx_ = this->get_parameter("camera_cx").as_double();
  camera_cy_ = this->get_parameter("camera_cy").as_double();
  camera_d_ = this->get_parameter("camera_d").as_double_array();

  // Create Publishers (Queue size = 1 to prevent latency buildup)
  image_pub_ = this->create_publisher<sensor_msgs::msg::Image>("/siyi/image_raw", 1);
  info_pub_ = this->create_publisher<sensor_msgs::msg::CameraInfo>("/siyi/camera_info", 1);

  // Build static CameraInfo msg
  camera_info_msg_ = build_camera_info();

  // Open RTSP
  open_capture();

  // Create Timer
  double timer_period_sec = 1.0 / target_fps_;
  timer_ = this->create_wall_timer(
    std::chrono::duration<double>(timer_period_sec),
    std::bind(&RtspPublisher::timer_callback, this)
  );

  RCLCPP_INFO(
    this->get_logger(),
    "SIYI RTSP Publisher C++ started: url=%s, flip_180=%s, target_fps=%.1f, resolution=%dx%d",
    rtsp_url_.c_str(), flip_180_ ? "true" : "false", target_fps_, image_width_, image_height_
  );
}

RtspPublisher::~RtspPublisher()
{
  if (cap_.isOpened()) {
    cap_.release();
    RCLCPP_INFO(this->get_logger(), "RTSP capture released");
  }
}

bool RtspPublisher::open_capture()
{
  if (cap_.isOpened()) {
    cap_.release();
  }

  RCLCPP_INFO(this->get_logger(), "Opening RTSP stream: %s", rtsp_url_.c_str());
  cap_.open(rtsp_url_, cv::CAP_FFMPEG);

  if (!cap_.isOpened()) {
    RCLCPP_ERROR(
      this->get_logger(),
      "Failed to open RTSP stream: %s. Will retry on next timer tick.",
      rtsp_url_.c_str()
    );
    return false;
  }

  cap_.set(cv::CAP_PROP_FRAME_WIDTH, image_width_);
  cap_.set(cv::CAP_PROP_FRAME_HEIGHT, image_height_);
  cap_.set(cv::CAP_PROP_BUFFERSIZE, 1);

  fail_count_ = 0;
  RCLCPP_INFO(this->get_logger(), "RTSP stream opened successfully");
  return true;
}

sensor_msgs::msg::CameraInfo RtspPublisher::build_camera_info()
{
  sensor_msgs::msg::CameraInfo msg;
  msg.header.frame_id = frame_id_;
  msg.width = image_width_;
  msg.height = image_height_;
  msg.distortion_model = "plumb_bob";
  msg.d = camera_d_;

  msg.k = {
    camera_fx_, 0.0,        camera_cx_,
    0.0,        camera_fy_, camera_cy_,
    0.0,        0.0,        1.0
  };

  msg.r = {
    1.0, 0.0, 0.0,
    0.0, 1.0, 0.0,
    0.0, 0.0, 1.0
  };

  msg.p = {
    camera_fx_, 0.0,        camera_cx_, 0.0,
    0.0,        camera_fy_, camera_cy_, 0.0,
    0.0,        0.0,        1.0,        0.0
  };

  return msg;
}

void RtspPublisher::timer_callback()
{
  if (!cap_.isOpened()) {
    fail_count_++;
    if (fail_count_ % 30 == 1) {
      RCLCPP_WARN(
        this->get_logger(),
        "RTSP not open, attempting reconnect... (fail_count=%d)",
        fail_count_
      );
    }
    open_capture();
    return;
  }

  cv::Mat frame;
  bool ret = cap_.read(frame);

  if (!ret || frame.empty()) {
    fail_count_++;
    if (fail_count_ >= max_consecutive_fails_) {
      RCLCPP_WARN(
        this->get_logger(),
        "Lost RTSP stream after %d failures, reconnecting...",
        fail_count_
      );
      open_capture();
    }
    return;
  }

  fail_count_ = 0;
  frame_count_++;

  // Resize frame if the physical stream size differs from requested parameters
  if (frame.cols != image_width_ || frame.rows != image_height_) {
    cv::resize(frame, frame, cv::Size(image_width_, image_height_));
  }

  if (flip_180_) {
    cv::flip(frame, frame, -1);
  }

  auto stamp = this->get_clock()->now();

  // Publish Image
  try {
    std_msgs::msg::Header header;
    header.stamp = stamp;
    header.frame_id = frame_id_;
    auto img_msg = cv_bridge::CvImage(header, "bgr8", frame).toImageMsg();
    image_pub_->publish(*img_msg);
  } catch (const std::exception & e) {
    RCLCPP_ERROR(this->get_logger(), "cv_bridge error: %s", e.what());
    return;
  }

  // Publish CameraInfo
  camera_info_msg_.header.stamp = stamp;
  info_pub_->publish(camera_info_msg_);

  // Log stats periodically (every 5 seconds)
  if (frame_count_ % (static_cast<int>(target_fps_) * 5) == 0) {
    RCLCPP_INFO(
      this->get_logger(),
      "Published %d frames (%dx%d) flip_180=%s",
      frame_count_, frame.cols, frame.rows, flip_180_ ? "true" : "false"
    );
  }
}

}  // namespace precision_landing

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(precision_landing::RtspPublisher)
